### PR TITLE
Add option to preserve URDF link ordering

### PR DESCRIPTION
**Pull request by aclegg3**
## Motivation and Context

For consistency with other simulators (e.g. pybullet), this PR adds an option to preserve the order of links parsed from URDF files as link indices.

To use this option:
`add_articulated_object_from_urdf(urdf_filename, maintain_link_order=true)`

## How Has This Been Tested

Added a pytest.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/esp/physics/bullet/BulletPhysicsManager.cpp [hunk 1]
```diff
@@ -162,20 +167,20 @@ int BulletPhysicsManager::addArticulatedObjectFromURDF(
   }
 
   // attach link visual shapes
-  int urdfLinkIx = 0;
-  for (auto& link : urdfImporter_->getModel()->m_links) {
-    if (link.second->m_visualArray.size() > 0) {
+  for (size_t urdfLinkIx = 0; urdfLinkIx < u2b->getModel()->m_links.size();
+       ++urdfLinkIx) {
+    auto urdfLink = u2b->getModel()->getLink(urdfLinkIx);
+    if (urdfLink->m_visualArray.size() > 0) {
```
Comment (Skylion007) [inline src/esp/physics/bullet/BulletPhysicsManager.cpp:173]:
```suggestion
    if (!urdfLink->m_visualArray.empty()) {
```

**Comment by dhruvbatra** [comment]
CC: @fmeier @akshararai

File: src/esp/physics/bullet/BulletURDFImporter.cpp [hunk 2]
```diff
@@ -197,6 +181,30 @@ void BulletURDFImporter::computeTotalNumberOfJoints(int linkIndex) {
...
+  cp.m_index = urdfLinkIndex;
+  cp.m_link_name = activeModel_->getLink(urdfLinkIndex)->m_name;
+  int mbIndex = cache->getMbIndexFromUrdfIndex(urdfLinkIndex);
+  cp.m_mbIndex = mbIndex;
+  cp.m_parentIndex = parentIndex;
+  int parentMbIndex =
+      parentIndex >= 0 ? cache->getMbIndexFromUrdfIndex(parentIndex) : -1;
+  cp.m_parentMBIndex = parentMbIndex;
+
+  allIndices.push_back(cp);
```
Comment (Skylion007) [inline src/esp/physics/bullet/BulletURDFImporter.cpp:198]:
```suggestion
  allIndices.push_back(std::move(cp));
```

File: src/esp/physics/bullet/BulletArticulatedObject.cpp [hunk 3]
```diff
@@ -85,10 +85,39 @@ void BulletArticulatedObject::initializeFromURDF(
...
+    parentTransforms[urdfLinkIndex] = rootTransformInWorldSpace;
+    std::vector<childParentIndex> allIndices;
+
+    u2b.getAllIndices(urdfLinkIndex, -1, allIndices);
+    std::sort(allIndices.begin(), allIndices.end(),
+              [](const childParentIndex& a, const childParentIndex& b) {
+                return a.m_index < b.m_index;
+              });
+
+    if (allIndices.size() + 1 >= parentTransforms.size()) {
```
Comment (jturner65) [inline src/esp/physics/bullet/BulletArticulatedObject.cpp:107]: Probably don't want to resize if they are equal, right?

**Review by jturner65 — APPROVED** [review]
LGTM.  We need to think where in the config ecosystem we would put this new field.

**Review by eundersander — APPROVED** [review]
See my comment but otherwise LGTM

File: src/esp/physics/PhysicsManager.h [hunk 4]
```diff
@@ -505,6 +505,8 @@ class PhysicsManager : public std::enable_shared_from_this<PhysicsManager> {
    * the components of the @ref ArticulatedObject.
    * @param forceReload If true, reload the source URDF from file, replacing the
    * cached model.
+   * @param maintainLinkOrder If true, maintain the order of link definitions
```
[the remark this answers is not in the document]
Comment (aclegg3) [reply src/esp/physics/PhysicsManager.h:508]: 🤷  Recursive traversal is more elegant for tree structures. Individual choice probably depends on whether you prefer consistency with respect to order of definition in the URDF file or some flattened preservation of parent->child relationships.

**Pull request merged**